api/src/utils/get-schema.ts

```diff
@@ -31,33 +33,44 @@ export async function getSchema(options?: {
 	const database = options?.database || getDatabase();
 	const schemaInspector = createInspector(database);
 
-	let result: SchemaOverview;
+	if (options?.bypassCache || env['CACHE_SCHEMA'] === false) {
+		return await getDatabaseSchema(database, schemaInspector);
+	}
 
-	if (!options?.bypassCache && env['CACHE_SCHEMA'] !== false) {
-		let cachedSchema;
+	const cached = await getSchemaCache();
 
-		try {
-			cachedSchema = await getSchemaCache();
-		} catch (err: any) {
-			logger.warn(err, `[schema-cache] Couldn't retrieve cache. ${err}`);
-		}
+	if (cached) {
+		return cached;
+	}
 
-		if (cachedSchema) {
-			result = cachedSchema;
-		} else {
-			result = await getDatabaseSchema(database, schemaInspector);
+	const lock = useLock();
+	const bus = useBus();
 
-			try {
-				await setSchemaCache(result);
-			} catch (err: any) {
-				logger.warn(err, `[schema-cache] Couldn't save cache. ${err}`);
-			}
-		}
-	} else {
-		result = await getDatabaseSchema(database, schemaInspector);
+	const lockKey = 'schemaCache--preparing';
+	const messageKey = 'schemaCache--done';
+	const processId = await lock.increment(lockKey);
+
+	const currentProcessShouldHandleOperation = processId === 1;
+
+	if (currentProcessShouldHandleOperation === false) {
+		logger.trace('Schema cache is prepared elsewhere, waiting for result.');
+
+		return new Promise((resolve) => {
+			bus.subscribe(messageKey, async () => {
+				const schema = await getSchema(options);
+				resolve(schema);
+			});
+		});
 	}
 
-	return result;
+	try {
+		const schema = await getDatabaseSchema(database, schemaInspector);
+		await setSchemaCache(schema);
+		return schema;
+	} finally {
+		bus.publish(messageKey, { ready: true });
+		await lock.delete(lockKey);
```

Comment:
This order feels a bit safer, so it should be guaranteed that the lock can be obtained again from a process receiving the `schemaCache--done` message
```suggestion
		await lock.delete(lockKey);
		bus.publish(messageKey, { ready: true });
```